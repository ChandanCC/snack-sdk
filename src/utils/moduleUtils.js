/* @flow */

import trim from 'lodash/trim';
import { types, parse, print } from 'recast';
import * as babylon from 'babylon';
import config from '../configs/babylon';

const parser = {
  parse: (code: string) => babylon.parse(code, config),
};

const b = types.builders;

const getModuleNameFromRequire = (node: *) => {
  const { callee, arguments: args } = node;

  let name;

  if (
    callee.name === 'require' &&
    args.length === 1 &&
    (args[0].type === 'StringLiteral' || args[0].type === 'TemplateLiteral')
  ) {
    if (args[0].type === 'StringLiteral') {
      name = args[0].value;
    } else if (args[0].type === 'TemplateLiteral' && args[0].quasis.length === 1) {
      name = args[0].quasis[0].value.cooked;
    }
  }

  return name && !/\n/.test(name) && !name.startsWith('/') ? name : null;
};

const getVersionFromComments = (comments: *) => {
  // Makes sure version specified is valid
  const versionFormat = /^(\s*(\d+\.)?(\d+\.)?(\*|\d+))|(LATEST)$/;

  return comments && versionFormat.test(comments[0].value)
    ? trim(comments[0].value.replace(/\s*/, ''))
    : null;
};

const addCommentToPath = (path, comment: string) => {
  const source = print(path).code;
  const lineComment = `// ${comment}`;
  const blockComment = `/* ${comment} */`;

  if (source.endsWith(lineComment) || source.endsWith(blockComment)) {
    return;
  }

  path.replace(parse(`${source} ${lineComment}`, { parser }).program.body[0]);
};

export const findModuleDependencies = (code: string): { [string]: string } => {
  const ast = parse(code, { parser });
  const dependencies: { [string]: string } = {};

  types.visit(ast, {
    visitImportDeclaration(path) {
      dependencies[path.node.source.value] = getVersionFromComments(path.node.trailingComments);

      this.traverse(path);
    },

    visitExportNamedDeclaration(path) {
      if (path.node.source) {
        dependencies[path.node.source.value] = getVersionFromComments(path.node.trailingComments);
      }
      this.traverse(path);
    },

    visitExportAllDeclaration(path) {
      dependencies[path.node.source.value] = getVersionFromComments(path.node.trailingComments);
      this.traverse(path);
    },

    visitCallExpression(path) {
      const name = getModuleNameFromRequire(path.node);

      if (name) {
        const { arguments: args } = path.node;

        // Comment location changes if user uses semicolons or not
        const version = getVersionFromComments(
          args[0].trailingComments || path.parentPath.parentPath.node.trailingComments
        );

        dependencies[name] = version;
      }

      this.traverse(path);
    },
  });

  return dependencies;
};

// Writes version number in comments in code
export const writeModuleVersions = (code: string, dependencies: { [string]: string }): string => {
  const ast = parse(code, { parser });

  types.visit(ast, {
    visitImportDeclaration(path) {
      const name = path.node.source.value;
      const version = dependencies[name];

      if (version) {
        addCommentToPath(path, version);
      }

      return false;
    },

    visitExportNamedDeclaration(path) {
      const name = path.node.source && path.node.source.value;
      const version = dependencies[name];

      if (version) {
        addCommentToPath(path, version);
      }

      return false;
    },

    visitExportAllDeclaration(path) {
      const name = path.node.source.value;
      const version = dependencies[name];

      if (version) {
        addCommentToPath(path, version);
      }

      return false;
    },

    visitCallExpression(path) {
      const name = getModuleNameFromRequire(path.node);

      if (name) {
        const version = dependencies[name];

        if (version) {
          addCommentToPath(path, version);
        }
      }

      return false;
    },
  });

  return print(ast).code;
};

