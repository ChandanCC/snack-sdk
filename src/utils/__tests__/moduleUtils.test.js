/* eslint-env jest */

import { parse, print } from 'recast';
import * as babylon from 'babylon';
import config from '../../configs/babylon';
import { findModuleDependencies, writeModuleVersions } from '../moduleUtils';

const parser = {
  parse: (code: string) => babylon.parse(code, config),
};

it('finds all imported modules', () => {
  const code = `
    import base64 from 'base64'; // 1.2.3
    import debounce from 'lodash/debounce'; /* 2.3.4 */
    import { connect } from 'react-redux';
  `;

  const dependencies = findModuleDependencies(code);

  expect(dependencies).toEqual({
    base64: '1.2.3',
    'lodash/debounce': '2.3.4',
    'react-redux': null,
  });
});

it('finds all required modules', () => {
  const code = `
    const base64 = require('base64'); // 1.2.3
    const debounce = require('lodash/debounce'); /* 2.3.4 */
    const { connect } = require('react-redux');
  `;

  const dependencies = findModuleDependencies(code);

  expect(dependencies).toEqual({
    base64: '1.2.3',
    'lodash/debounce': '2.3.4',
    'react-redux': null,
  });
});

it('finds all required modules with backticks', () => {
  const code = `
    const base64 = require(\`base64\`); // 1.2.3
    const debounce = require(\`lodash/debounce\`); /* 2.3.4 */
    const { connect } = require(\`react-redux\`);
  `;

  const dependencies = findModuleDependencies(code);

  expect(dependencies).toEqual({
    base64: '1.2.3',
    'lodash/debounce': '2.3.4',
    'react-redux': null,
  });
});

it('finds dependencies using all import styles', () => {
  const code = `
    import v from "mod1"; // 1.0.0
    import * as ns from "mod2"; // 2.0.0
    import {x} from "mod3"; // 3.0.0
    import {x as v} from "mod4"; // 4.0.0
    import "mod5"; // 5.0.0

    export {x} from "mod6"; // 6.0.0
    export {x as v} from "mod7"; // 7.0.0
    export * from "mod8"; // 8.0.0

    export default 7;
    export const value = 6;
    const otherValue = 5;
    export { otherValue }
  `;

  const dependencies = findModuleDependencies(code);
  expect(dependencies).toEqual({
    mod1: '1.0.0',
    mod2: '2.0.0',
    mod3: '3.0.0',
    mod4: '4.0.0',
    mod5: '5.0.0',
    mod6: '6.0.0',
    mod7: '7.0.0',
    mod8: '8.0.0',
  });
});

it('writes dependency pins for all import styles', () => {
  const code = `
    import v from "mod1";
    import * as ns from "mod2";
    import {x} from "mod3";
    import {x as v} from "mod4";
    import "mod5";

    export {x} from "mod6";
    export {x as v} from "mod7";
    export * from "mod8";

    export default 7;
    export const value = 6;
    const otherValue = 5;
    export { otherValue }

  `;
  const modules = {
    mod1: { version: '1.0.0' },
    mod2: { version: '2.0.0' },
    mod3: { version: '3.0.0' },
    mod4: { version: '4.0.0' },
    mod5: { version: '5.0.0' },
    mod6: { version: '6.0.0' },
    mod7: { version: '7.0.0' },
    mod8: { version: '8.0.0' },
  };

  const result = writeModuleVersions(code, modules);
  expect(result).toMatchSnapshot();
});

it('replaces dependency pins for all import styles', () => {
  const code = `
    import v from "mod1"; // 2.3.1
    import * as ns from "mod2"; // 4.3.2
    import {x} from "mod3"; // 5.4.3
    import {x as v} from "mod4"; // 6.5.4
    import "mod5"; // 7.6.5

    export {x} from "mod6"; // 8.7.6
    export {x as v} from "mod7"; // 9.8.7
    export * from "mod8"; // 1.2.3

    export default 7; // 2.3.4
    export const value = 6; // 3.4.5
    const otherValue = 5; // 5.7.6
    export { otherValue } // 6.7.8

  `;
  const modules = {
    mod1: { version: '1.0.0' },
    mod2: { version: '2.0.0' },
    mod3: { version: '3.0.0' },
    mod4: { version: '4.0.0' },
    mod5: { version: '5.0.0' },
    mod6: { version: '6.0.0' },
    mod7: { version: '7.0.0' },
    mod8: { version: '8.0.0' },
  };

  const result = writeModuleVersions(code, modules);
  expect(result).toMatchSnapshot();
});

it("doesn't parse non-static and invalid requires", () => {
  const code = `
    const base64 = require();
    const debounce = require('debounce', null); // 2.3.4
    const moment = require(\`\${name}\`);
    const leftpad = require(\`left-\${name}\`);
    const core = require('/core');
    const promise = require('promise\\nme');
    const bluebird = require(\`blue
      bird
    \`);
    const { connect } = require(10);
  `;

  const dependencies = findModuleDependencies(code);

  expect(dependencies).toEqual({});
});

it('writes versions for imported modules', () => {
  const code = `
    import base64 from 'base64';
    import debounce from 'lodash/debounce';
    import { connect } from 'react-redux';
  `;

  const modules = {
    base64: { version: '1.2.3' },
    'lodash/debounce': { version: '2.3.4' },
  };

  const result = writeModuleVersions(code, modules);

  expect(result).toMatchSnapshot();
});

it('writes versions for required modules', () => {
  const code = `
    const base64 = require('base64');
    const debounce = require('lodash/debounce');
    const { connect } = require('react-redux');
  `;

  const modules = {
    base64: { version: '1.2.3' },
    'lodash/debounce': { version: '2.3.4' },
  };

  const result = writeModuleVersions(code, modules);

  expect(result).toMatchSnapshot();
});

it('writes versions for required modules with backticks', () => {
  const code = `
    const base64 = require(\`base64\`);
    const debounce = require(\`lodash/debounce\`);
    const { connect } = require(\`react-redux\`);
  `;

  const modules = {
    base64: { version: '1.2.3' },
    'lodash/debounce': { version: '2.3.4' },
  };

  const result = writeModuleVersions(code, modules);

  expect(result).toMatchSnapshot();
});
