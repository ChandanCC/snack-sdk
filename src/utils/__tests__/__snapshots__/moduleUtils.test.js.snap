// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`writes dependency pins for all import styles 1`] = `
"
    import v from \\"mod1\\"; /* 1.0.0 */
    import * as ns from \\"mod2\\"; /* 2.0.0 */
    import {x} from \\"mod3\\"; /* 3.0.0 */
    import {x as v} from \\"mod4\\"; /* 4.0.0 */
    import \\"mod5\\"; /* 5.0.0 */

    export {x} from \\"mod6\\"; /* 6.0.0 */
    export {x as v} from \\"mod7\\"; /* 7.0.0 */
    export * from \\"mod8\\"; /* 8.0.0 */

    export default 7;
    export const value = 6;
    const otherValue = 5;
    export { otherValue }

  "
`;

exports[`writes versions for imported modules 1`] = `
"
    import base64 from 'base64'; /* 1.2.3 */
    import debounce from 'lodash/debounce'; /* 2.3.4 */
    import { connect } from 'react-redux';
  "
`;

exports[`writes versions for required modules 1`] = `
"
    const base64 = require('base64') /* 1.2.3 */;
    const debounce = require('lodash/debounce') /* 2.3.4 */;
    const { connect } = require('react-redux');
  "
`;

exports[`writes versions for required modules with backticks 1`] = `
"
    const base64 = require(\`base64\`) /* 1.2.3 */;
    const debounce = require(\`lodash/debounce\`) /* 2.3.4 */;
    const { connect } = require(\`react-redux\`);
  "
`;
